closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -469,39 +476,52 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
+        int bufSize = 0;
+        byte[] keyCopy;
+        byte[] ivCopy;
+        byte[] aad;
+
+        // need to synchronize this block as if there are concurrent calls to decryptFinal
+        // the shared variables aadBuffer and ibuffer may corrupt input and cause a segfault
+        // in openssl code
         if (len < 0) {
             throw new ProviderException("Input length is negative");
         }
-        if (len < tagLenBytes - ibuffer.size()) {
-            throw new AEADBadTagException("Input too short - need tag");
-        }
-        if (len > MAX_BUF_SIZE - ibuffer.size()) {
-            throw new ProviderException("SunJCE provider only supports "
-                + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
-        }
 
-        if (out.length - outOfs < len  + ibuffer.size() - tagLenBytes) {
-            throw new ShortBufferException("Output buffer too small");
-        }
+        synchronized (this) {
+            bufSize = ibuffer.size();
+            if (len > 0) {
+                ibuffer.write(in, inOfs, len);
+            }
 
-        byte[] aad = ((aadBuffer == null || aadBuffer.size() == 0) ?
-                       emptyAAD : aadBuffer.toByteArray());
+            if (len < (tagLenBytes - bufSize)) {
```

Comment:
A local copy of `tagLenBytes` should be created so that value can be used after the synchronized block.

Reply:
I added a local copy of `tagLenBytes`

Reply:
It should be read within the `synchronized` block.